### ISSUE TITLE
Ignoring URI outside base dir

### ISSUE THREAD

**Issue by oliverjkb** [mopidy/mopidy#1268]
Don't really want to make this an issue, but I thought this might be the best place to solve it. 

I just switched to Local-SQLite because I didn't want to restart Mopidy after adding new music files. 
Now the local scan turns out to work just fine, but when I try to play any of my last.fm test files, music will not be played and the mopidy log shows the following block of warnings: 
```
2015-08-26 16:15:22,618 WARNING [19255:FileBackend-5] mopidy.file.library: Ignoring URI outside base dir: /media/disk/samba/music/Com Truise - Cyanide Sisters.mp3
2015-08-26 16:15:22,619 WARNING [19255:FileBackend-5] mopidy.file.library: Ignoring URI outside base dir: /media/disk/samba/music/Death Grips - Beware.mp3
2015-08-26 16:15:22,619 WARNING [19255:FileBackend-5] mopidy.file.library: Ignoring URI outside base dir: /media/disk/samba/music/Death Grips - I've Seen Footage.mp3
2015-08-26 16:15:22,619 WARNING [19255:FileBackend-5] mopidy.file.library: Ignoring URI outside base dir: /media/disk/samba/music/Death Grips - No Love.mp3
2015-08-26 16:15:22,620 WARNING [19255:FileBackend-5] mopidy.file.library: Ignoring URI outside base dir: /media/disk/samba/music/Sleepmakeswaves - It's Dark, It's Cold, It's Winter.mp3
```

in mopidy.conf I have the following configuration set:

```
[local]
enabled = true
library = sqlite
media_dir = /media/disk/samba/music
```
Unfortunately I don't have a clue, where the Problem might come from. 
The only things I changed were to set "library" so sqlite and to add a block 
```
[local-sqlite]
enabled = true
```
which end up to shown as the full config block in when I do a "mopidyctl config" : 

```
[local-sqlite]
enabled = true
directories = 
  Albums                  local:directory?type=album
  Artists                 local:directory?type=artist
  Composers               local:directory?type=artist&role=composer
  Folders                 local:directory:
  Genres                  local:directory?type=genre
  Performers              local:directory?type=artist&role=performer
  Release Years           local:directory?type=date&format=%25Y
  Tracks                  local:directory?type=track
  Last Week's Updates     local:directory?max-age=604800
  Last Month's Updates    local:directory?max-age=2592000
timeout = 10
use_album_mbid_uri = true
use_artist_mbid_uri = false
search_limit = -1
```
Is it possible that "local:directory?" is causing the error? If so, how can I prevent it from using all the directories?

**Comment by kingosticks** [comment]
The best place for something like this is http://mopidy.discuss.com. 

Are you aware that the File backend is not the same as the Local backend?

**Comment by tkem** [comment]
@kingosticks: Although you're basically right about `mopidy.discuss.com`, this *might* really be an issue with Mopidy itself. IIRC (it's been more than three weeks since I last worked on `sqlite`, which is my maximum attention/memory span nowadays), the local playback part will transform `local` URIs to `file` URIs. Now, since `file` is an extension of it's own, may it be the case that the `file` backend groaks if a `file` URI lies outside its configured directories? Are `local` and `file` dirs synchronized in any way? Why does the `file` backend have "allowed" dirs, in the first place?

**Comment by tkem** [comment]
@oliverjkb: You might try adding `/media/disk/samba/music` to `file/media_dirs`; haven't checked that myself (no access to Mopidy currently @work), but I think this might work. See also the `file` docs at https://docs.mopidy.com/en/latest/ext/file/
Please teel us if that worked; if it does, we might want to do something about it ;-)

**Comment by kingosticks** [comment]
The File backend only plays files that lie within the configured `file/media_dirs` by design.

**Comment by tkem** [comment]
@kingosticks: Designs should be questioned from time to time ;-)
And I'm pretty sure nobody thought about how `local/media_dir` interacts with `file/media_dirs` until now. At least, I didn't. So I'd say either let `file` play any URI I throw at it if it can (my +1), or somehow merge `local/media_dir` with `file/media_dirs`.

**Comment by kingosticks** [comment]
True, and I can't find any rationale for it now.

**Comment by tkem** [comment]
@jodal, @adamcik: I suspect `FileLibraryProvider.lookup()` checking if a URI is within configured `media_dirs` to be the cause of this, but after looking at the `mopidy.audio` code for some time, my brain hurts. Could you take a look? And if so, is there any good reason why `file/media_dirs` limit access to `file` URIs beyond browsing?

**Comment by jodal** [comment]
I think @tkem's analysis is correct, and we should remove the check in `FileLibraryProvider.lookup()`.

**Comment by adamcik** [comment]
Having the check in `browse` was intentional to limit browsing your files. Personally I was considering not even having `$HOME` in there. But having the limitation in `lookup` doesn't make as much sense. And yes if mopidy-local-sqlite returns file URIs directly then that is very likely what is happening.

**Comment by tkem** [comment]
@adamcik: Well, it's not Mopidy-Local-SQLite that returns `file` URIs, it's the `mopidy.local.LocalPlaybackProvider.translate_uri()` that does the translation, and I got lost somewhere in between...

**Comment by adamcik** [comment]
Hmm, that is weird. Things coming out of `translate_uri` should only ever make it to GStreamer, not any of our code that would do this checking.

**Comment by tkem** [comment]
Yeah, that's why I got confused and asked the powers that be for help ;-)

**Comment by tkem** [comment]
@oliverjkb: It would be great if you could try to switch back to the `json` library and see if that problem perists. This would help us to narrow down the source of that issue...

**Comment by oliverjkb** [comment]
I'll not have access to that machine for a couple of days now. Non the less I think I'll be able to reproduce that behaviour tomorrow. Will come back to that issue asap. Promise!

**Comment by tkem** [comment]
@oliverjkb: With Mopidy-Local-SQLite installed, did you browse to the files via "Local media -> Folders"? If so, there is indeed an issue with Mopidy-Local-SQLite and Mopidy v1.1; the "virtual" directories (Artists, Albums, Genres) should work fine, and search should work too, AFAICS.
Also, you shouldn't have this issue with the "json" library.
 
When doing "Folders" browsing, which is basically what the new `file` extension does, Mopidy-Local-SQLite returns `file` URIs. This of course conflicts with the `file` settings, and leads to a failed `lookup` call. Now, the whole "Folders" stuff could be removed from the sqlite extension, since it's only duplicated functionality. Or the `file` lookup could be less picky about allowed paths.

**Comment by tkem** [comment]
BTW, the reason SQLite uses `file` URIs here is that I wanted to browse/lookup/play files that weren't indexed yet via `mopidy local scan` ;-)

**Comment by jodal** [comment]
I can reproduce this with Mopidy-Local-SQLite, but not with Mopidy-Local with the JSON library. That probably because -SQLite use the `file:` scheme on files it is browsing, while the JSON library using `local:` for the files, as @tkem explained.

It all works when removing the is-within-file/media_dirs check in `FileLibrary.lookup()`.

**Comment by oliverjkb** [comment]
Hi and sorry for not getting my hands on the keyboard earlier. Work got me blocked longer than I wished it would. 
Indeed, using Mopidy-Local, hence updating the json library, works like a charm. No Problems whatsoever. 
If there's anything I could try or test, I'd be glad doing so.

**Comment by jodal** [comment]
A fix has been committed to the release-1.1 branch of Mopidy, and will be released as 1.1.1 soon.

**Comment by oliverjkb** [comment]
@jodal :+1:  Great!
